File: kangtaku/triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size(), i, j;
        int[][] d = new int[2][n];
        d[0][0] = triangle.get(0).get(0);
        for (i = 1; i < n; i++) {
            for (j = 0; j <= i; j++) {
                int min = Integer.MAX_VALUE, now = triangle.get(i).get(j);
                if (j > 0 && min > now + d[(i - 1) % 2][j - 1]) {
                    min = now + d[(i - 1) % 2][j - 1];
                }
                if (j < i && min > now + d[(i - 1) % 2][j]) {
                    min = now + d[(i - 1) % 2][j];
                }
                d[i % 2][j] = min;
                //System.out.print(min + " ");
            }
            //System.out.println();
        }
        int ans = Integer.MAX_VALUE;
        for (i = 0; i < n; i++) {
            ans = Math.min(ans, d[(n - 1) % 2][i]);
        }
        return ans;
    }
}

File: iamminji/leetcode/medium/triangle.py
# 120. Triangle
# https://leetcode.com/problems/triangle/description/


class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """

        arr = triangle[-1]

        length = len(triangle[-1])
        if length == 1:
            return triangle[0][0]
        elif length == 2:
            return min(triangle[0][0] + triangle[1][1], triangle[0][0] + triangle[1][0])

        for idx in range(len(triangle) - 2, -1, -1):
            sub = triangle[idx]
            jdx = 0
            while jdx < length - 1:
                if jdx == 0:
                    arr[jdx] = min(sub[jdx] + arr[jdx], sub[jdx] + arr[jdx + 1])
                    jdx += 1
                    continue
                arr[jdx - 1] = min(sub[jdx - 1] + arr[jdx], arr[jdx - 1])
                arr[jdx] = min(sub[jdx] + arr[jdx], sub[jdx] + arr[jdx + 1])
                jdx += 1
            length = len(sub)

        return arr[0]


if __name__ == "__main__":
    sol = Solution()
    print sol.minimumTotal([[-1], [2, 3], [1, -1, -3]])
